gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/x509/X509AuthenticationUserDetailsService.groovy

```diff
@@ -56,7 +59,14 @@ class X509AuthenticationUserDetailsService implements AuthenticationUserDetailsS
 
     def x509 = (X509Certificate) token.credentials
     def email = emailFromSubjectAlternativeName(x509) ?: token.principal
-    def roles = userRolesProvider.loadRoles(email as String)
+
+    def roles
+    if (rolesExtractor) {
+      roles = rolesExtractor.fromCertificate(x509).orElse(userRolesProvider.loadRoles(email as String))
```

Comment:
Totally fair. 

In our case I suspect we'll be happy with only the groups from the certificate but no harm in merging.